### ISSUE TITLE
Messages are unhashable

### ISSUE THREAD

**Issue by jasper-lyons** [olemb/mido#36]
This means they can't meaningfully be used as keys in dicts, put in sets etc.

I'm building a Markov Model for generating drum midi and being able to hash Messages would make my life 10X easier. I'm happy to put code it up and submit the pull request I just wanted to ask if:

a) This fits in with your plans for Mido?
b) You have enough time to merge the pull request?
c) There are any special requirements for contributing not already mentioned?

**Comment by olemb** [comment]
Thanks for writing. First, to answer your questions:

a) This fits very well with my plans for Mido if there is a good way do to implement it. See below.
b) I will find the time.
c) There are no special requirements for contributing.

I would love to make messages hashable. Unfortunately this would be problematic as they are mutable. Consider the following code:

```
from mido import Message

# Monkey patch messages to be hashable. (Don't do this.)
Message.__hash__ = lambda msg: hash(tuple(msg.__dict__.items()))

msg = Message('clock', time=0)

# Add message twice:
unique = set()
unique.add(msg)
msg.time = 1
unique.add(msg)

for msg in unique:
    print(msg)
```

This produces a set containing the same message twice:

```
clock time=1
clock time=1
```

I made two mistakes when I designed messages:

**Mistake 1: They are mutable.**

```
This causes a lot of problems. For example when you send a message the port has to make a copy of it to ensure that the sender is free to change their copy later. Making messages immutable would make it a lot easier to pass messages around (which is what they're for), but changing it now would probably cause a lot of breakage. I am thinking about adding frozen messages so you could do:

>>> from mido import Message, FrozenMessage
>>> msg1 = FrozenMessage('clock', time=0)
>>> msg2 = freeze(Message('clock', time=0))
>>> unique = {msg1, msg2}
>>> len(unique)
1
```

This should be relatively easy to implement but also very clunky. You'd have to use `freeze()` everywhere and get in trouble if if you didn't.

**Mistake 2: Message comparison doesn't include time.**

This makes it very hard to compare tracks. You can't just do 'track1 == track2' as that would ignore the time attributes. I am seriously considering changing this even if it breaks code. For two messages with the different time stamps to be the same doesn't make any sense at all:

```
>>> from mido import Message
>>> msg1 = Message('clock', time=0)
>>> msg2 = Message('clock', time=1)
>>> msg1 == msg2
True
>>> str(msg1) == str(msg2)
False
>>> repr(msg1) == repr(msg2)
False
```

I hope there's a good way to implement this. If not, maybe it's time for a (somewhat) backwards incompatible Mido 2?

**Comment by olemb** [comment]
Whatever the case in the long term I will help you find something that you can use right now.

**Comment by jasper-lyons** [comment]
Thanks. Best response I've ever received from a project maintainer.

I am with you on your solution for 1), that would make my use case super simple.

A midi message one second in the future could be the same message as one sent now so not having time in the comparison makes sense though is not amazingly practical.

One approach might be to abstract time from messages and pop it into tracks, perhaps as some sort of key for messages / groups of messages?

**Comment by olemb** [comment]
I think the only thing that makes sense is to include time in comparisons. It's too confusing and inconsistent the way it is today. I am worried about breaking code but I'm not sure how many people actually rely on the current behaviour. I would love to hear from more people on this before I make a decision.

I have though about this a lot today and I've had some ideas and made some decisions.

First off, I've launched Project Elsa, an experimental branch which adds frozen messages to Mido. I think I've figured out how to implement them in a simple, straight forward and 100% backwards compatible way. I will have some code up later this weekend. Feel free to contribute code or ideas. The branch is called ``elsa``.

I've started another branch where message comparison includes time. This branch is called ``compare-time``.

I've also started making plans for Mido 2.0. I will put up a repository with some ideas and experimental code soon.

I have a lot of ideas for Mido 2.0 but I would like to include everyone in the planning and implementing so I need to find a good place to have these discussions. Let me know if you have any ideas.

**Comment by olemb** [comment]
Another option is to leave comparison the way it is but include time when comparing tracks. Two tracks with the same messages but different times are not equal.

**Comment by olemb** [comment]
Frozen messages are a thing. Check out the "elsa" branch.

I haven't written any documentation yet, but here are the basics:

```
>>> from mido import Message, FrozenMessage
>>> msg = Message('clock')
>>> fmsg = FrozenMessage('clock')

>>> msg
<message clock time=0>
>>> fmsg
<frozen message clock time=0>

>>> msg == f
True

>>> msg.frozen
False
>>> fmsg.frozen
True

>>> hash(msg)
ValueError: only frozen messages are hashable
>>> fmsg.time = 1
ValueError: frozen messages are immutable

>>> hash(fmsg)
-3995925068487577597
>>> s = set()
>>> s.add(fmsg)
>>> s.add(fmsg)
>>> s
{<frozen message clock time=0>}
```

You can convert back and forth:

```
>>> FrozenMessage(msg)
<frozen message clock>
>>> Message(fmsg)
<message clock time=0>
>>> Message(fmsg, time=1)
<message clock time=1>
```

(``Message(msg)`` can now be used as an alternative to ``msg.copy()``.)

Meta messages work the same way. Ports and parsers will take the parameter ``emit_frozen`` (not implemented yet):

```
>>> port = mido.open_input(emit_frozen=True)
>>> port.receive()
<frozen message note_on channel=0 note=0 velocity=64 time=0>
```

Hashing ignores time just like comparison. How will that work for your Markov model?

**Comment by olemb** [comment]
I've renamed the branch to a more neutral ``frozen-messages``.

**Comment by olemb** [comment]
I've implemented frozen messages in the new experimental ``frozen-message`` branch.

You bring up some good poiints

**Comment by olemb** [comment]
(I sent you a longer reply but it was a bit all over the place so I've replaced it with this one I wrote when I was more awake. I see that I also accidentally closed the issue without realizing. I feel like I'm probably not living up to first impressions. Anyway...)

You make a couple of good points. One one hand a message in the future is the same as that message today. Also time is naturally a part of the track and not the message. On the other hand that time attribute is very useful.

What's clear to me is that it should be one or the other, not a little bit of each like today. If time is part of the message it should be part of the comparison and hashing as well.

I have an implementation of this in my ``compare-time`` branch.

I've also implemented frozen messages::

```
>>> from mido import Message, FrozenMessage
>>> msg = Message('clock')
>>> fmsg = FrozenMessage('clock')
>>> msg
<message clock time=0>
>>> fmsg
<frozen message clock time=0>
>>> fmsg.time = 1
ValueError: frozen messages are immutable
```

You can tell which is which by the ``frozen`` attribute::

```
>>> fmsg.frozen
True
```

and convert between the two like this::

```
>>> FrozenMessage(msg, time=1)
<frozen message clock time=1>
```

I'm very happy with how this turned out. I think a combination of the two patches would be the best solution.

**Comment by olemb** [comment]
Have you had a chance to try the frozen messages?

I'm a bit conflicted about where to take this. On once hand the frozen messages are really handy. On the other they complicate the library with a new concept and three new classes, one of which is a mixin. Ideally I'd like to make all messages immutable, but I'm not sure how well that would be received. I'd like to involve all Mido users in that decision.

So frozen messages are on, ehem, ice at the moment.

**Comment by jasper-lyons** [comment]
Unfortunately not yet! I will give them a play asap though.

Perhaps to save complexity in the current API and given the the current implementation is a mixin, the frozen implementation could exist in a separate package? Then, until such time as a v2.0 / release worthy updating the api (if ever), one could simply:

```python
from mido import MidiFile, Message
from mido_immutable import FrozenMessage, freeze
```

Then use the hash-able messages as required?

Just a thought.

**Comment by olemb** [comment]
Yes, that's a good idea. The ``msg.frozen`` attribute would have to be dropped though but I don't think that's a big deal.

**Comment by jasper-lyons** [comment]
No existing code is using it atm and python will just throw an error when we try to hash `Message`.

**Comment by olemb** [comment]
I don't know if you still working need them but I just wanted to let you know that I've added frozen messages. They are available in my develop branch and will be included in 1.2.0.

To use them:

```
from mido.frozen import FrozenMessage, FrozenMetaMessage
```

The API may change a bit before release. For now there is no `frozen` attribute and no `freeze` function and also no way to thaw frozen messages.

I've completely reimplemented messages for 1.2.0 which made this much easier to add.

**Comment by jasper-lyons** [comment]
Awesome! This will let me simplify some stuff greatly :)

**Comment by olemb** [comment]
I just released 1.2:

     https://mido.readthedocs.io/en/latest/frozen_messages.html

I guess I'll close the issue.

**Comment by jasper-lyons** [comment]
Looks great!
